### ISSUE TITLE
Proper way of applying ZCA

### ISSUE THREAD

**Issue by RaananHadar** [fchollet/keras#2947]
Hi,

I tried looking for examples on how to effectively apply ZCA in keras.
My first attempt to use ZCA did not work, I attempted to use it in a similar way to @asmith26 in [his work here](https://github.com/asmith26/wide_resnets_keras/blob/master/main.py).

``` python

    train_datagen = ImageDataGenerator(
                                featurewise_center=True,
                                featurewise_std_normalization=True,
                                zca_whitening=True,
                                horizontal_flip=True)
    train_datagen.fit(X_train, augment=True, rounds=2)

    test_datagen = ImageDataGenerator(
                     featurewise_center=True,
                     featurewise_std_normalization=True)
    test_datagen.fit(X_train)
```

This provides excellent accuracy on the training but fails spectacularly on the test data - probably because the test data does not apply the transform.
One can attempt to use the train_datagen.flow() on the test data but it is inconvenient: 
Often, one would not want to perform data augmentation on the test data and it results redundant computations. This also undesired in a production environment.

So, in summary, how would you apply ZCA on the training and the test without any data augmentation on the test?

Thank you!

**Comment by fchollet** [comment]
You can read the code, see what transformations ZCA whitening does, then reapply these transformations on your data. Your can also instantiate an `ImageDataGenerator` that will do it for you.

**Comment by MartinThoma** [comment]
Here is a complete example how to use ZCA of the ImageDataGenerator: http://stackoverflow.com/a/43069409/562769
